discovery_engine_core/core/src/engine.rs

```diff
@@ -1002,50 +934,58 @@ impl XaynAiEngine {
             Box::new(PersonalizedNews::new(&endpoint_config, client.clone())) as BoxedOps,
         ];
 
+        let (stack_data, builder) = if let Some(state) = state {
+            if stack_ops.is_empty() {
+                return Err(Error::NoStackOps);
+            }
+            State::deserialize(state, builder)?
+        } else {
+            (HashMap::default(), builder)
+        };
+        let ranker = builder.build()?;
         #[cfg(feature = "storage")]
         let storage = {
             let storage = SqliteStorage::connect("sqlite::memory:").await?;
             storage.init_database().await?;
             Box::new(storage) as _
         };
-        if let Some(state) = state {
-            let state: State = bincode::deserialize(state).map_err(Error::Deserialization)?;
-            let ranker = builder.with_serialized_state(&state.ranker.0)?.build()?;
-            Self::from_state(
-                &state.engine,
-                endpoint_config,
-                core_config,
-                ranker,
-                history,
-                stack_ops,
-                client,
-                #[cfg(feature = "storage")]
-                storage,
-            )
-            .await
-        } else {
-            let ranker = builder.build()?;
-            Self::new(
-                endpoint_config,
-                core_config,
-                ranker,
-                history,
-                stack_ops,
-                client,
-                #[cfg(feature = "storage")]
-                storage,
-            )
-            .await
-        }
+
+        Self::new(
+            endpoint_config,
+            core_config,
+            ranker,
+            history,
+            stack_data,
+            stack_ops,
+            client,
+            #[cfg(feature = "storage")]
+            storage,
+        )
+        .await
     }
 }
 
 #[derive(Serialize, Deserialize)]
 struct StackState(Vec<u8>);
 
+impl StackState {
+    fn deserialize(&self) -> Result<HashMap<StackId, StackData>, Error> {
+        bincode::deserialize(&self.0).map_err(Error::Deserialization)
+    }
+}
+
 #[derive(Serialize, Deserialize)]
 struct RankerState(Vec<u8>);
 
+impl RankerState {
+    fn deserialize<'a>(
+        &self,
+        builder: Builder<'a, AveragePooler>,
+    ) -> Result<Builder<'a, AveragePooler>, Error> {
+        builder.with_serialized_state(&self.0).map_err(Into::into)
```

Comment:
just a question for my benefit, could this be written like below and would it be more/less clear in your opinion? 

```rust
Ok(builder.with_serialized_state(&self.0)?)
```

Reply:
it could very well be written in the suggested way. the `?` operator internally uses some error mapping to return the desired error type.

regarding readability, i personally prefer the sequential functional style, but this is very subjective of course. that means, for me it's clearer to read the line as "use the builder, then call some serialized state method, then do some error mapping, then return". with the `Ok()` wrapper i have to step back and forth while reading the line, which makes it more difficult imo, but again this is subjective.

there are some unofficial design patterns [like this page](https://rust-unofficial.github.io/patterns/intro.html), but i didn't find if the is an idiomatic way for this kind of error handling.